src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Adapter\PhpAmqp;
+
+use Symfony\Component\Messenger\Adapter\PhpAmqp\Exception\RejectMessageExceptionInterface;
+use Symfony\Component\Messenger\Transport\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\DecoderInterface;
+
+/**
+ * Symfony Messenger receiver to get messages from AMQP brokers using PHP's AMQP extension.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class AmqpReceiver implements ReceiverInterface
+{
+    public const STOP = 'stop';
+
+    private $messageDecoder;
+    private $connection;
+
+    public function __construct(DecoderInterface $messageDecoder, Connection $connection)
+    {
+        $this->messageDecoder = $messageDecoder;
+        $this->connection = $connection;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(): iterable
+    {
+        $shouldStop = false;
+
+        while (!$shouldStop) {
+            if (null === $message = $this->connection->waitAndGet()) {
+                continue;
+            }
+
+            try {
+                $shouldStop = self::STOP === (yield $this->messageDecoder->decode(array(
+                    'body' => $message->getBody(),
+                    'headers' => $message->getHeaders(),
```

Comment:
Why don't you passe the full messages ? I could be needed to build a custom retry-strategy

Reply:
Because I would be AMQP-specific. The contract of the encoder/decoder is a plain PHP array, nothing more: this can support different type of messages. Though, I'm sure we could add much more of the message's properties to it if necessary (and I suggest we do this when adding the retry-strategy) in the next iteration :)